// decode-engine/specs/enfora.js
module.exports = [{field: "API_NUMBER",
    index: 0,
    expression: null,
    LEorBE : "BE",
    type : "smallInt",
    signed : false,
    count: 2},
    {field: "API_COMMAND_TYPE",
        index: 2,
        expression: null,
        LEorBE : "BE",
        type : "byteInt",
        signed : false,
        count: 1},
    {field: "API_OPT_HDR_LENGTH",
        index: 3,
        expression: null,
        LEorBE : "BE",
        type : "byteInt",
        signed : false,
        count: 1},
    {field: "eventNumber",
        index: 4,
        expression: null,
        LEorBE : "BE",
        type : "integer",
        signed : false,
        count: 4},
    {field: "modemUnitId",
        index: 8,
        expression: null,
        LEorBE : "BE",
        type : "longInt",
        signed : false,
        count: 22},
    {field: "GPIO",
        index: 30,
        expression: null,
        LEorBE : "BE",
        type : "smallInt",
        signed : false,
        count: 2},
    {field: "GPS_UTC_DATE",
        index: 32,
        expression: null,
        LEorBE : "BE",
        type : "integer",
        signed : false,
        count: 3},
    {field: "GPS_STATUS",
        index: 36,
        expression: null,
        type : "byteInt",
        signed : false,
        count: 1},
    {field: "latitude",
        index : 37,
        expression:function(input){
            var dd =  Math.round(input/100000);
            var mmsss =  (input%100000)/1000;
            return dd + (mmsss/60)},
        type : "integer",
        LEorBE : "BE",
        signed : true,
        count: 3},
    {field: "longitude",
        index : 40,
        expression: function(input){
            var dd =  Math.round(input/100000);
            var mmsss =  (input%100000)/1000;
            return dd + (mmsss/60)},
        LEorBE : "BE",
        type : "integer",
        signed : true,
        count: 4},
    {field: "speed",
        index: 44,
        expression: function(input){
            return (input * 0.115155)},
        type : "smallInt",
        signed : false,
        count: 2},
    {field: "heading",
        index: 46,
        expression: null,
        type : "smallInt",
        signed : false,
        count: 2},
    {field: "GPS_UTC_TIME",
        index: 48,
        expression: null,
        type : "integer",
        signed : false,
        count: 3},
    {field: "GPS_SAT_COUNT",
        index: 51,
        expression: null,
        type : "byteInt",
        signed : false,
        count: 1},
    {field: "odometer",
        index: 52,
        expression: null,
        type : "integer",
        signed : false,
        count: 4},
    {field: "year",
        index: 51,
        expression: null,
        type : "byteInt",
        signed : false,
        count: 1},
    {field: "month",
        index: 51,
        expression: null,
        type : "byteInt",
        signed : false,
        count: 1},
    {field: "day",
        index: 51,
        expression: null,
        type : "byteInt",
        signed : false,
        count: 1},
    {field: "hour",
        index: 51,
        expression: null,
        type : "byteInt",
        signed : false,
        count: 1},
    {field: "minute",
        index: 51,
        expression: null,
        type : "byteInt",
        signed : false,
        count: 1},
    {field: "second",
        index: 51,
        expression: null,
        type : "byteInt",
        signed : false,
        count: 1}]
