// decode-engine/specs/montage.js

module.exports= [
    {field: "reportTag",
        index: 0,
        expression: null,
        type : "byteInt",
        signed : false,
        count: 1},
    {field: "imei",
        index: 1,
        expression: null,
        signed : false,
        type : "string",
        stringType : 'hex',
        count: 8},
    {field: "imsi",
        index: 9,
        expression: null,
        signed : false,
        type : "string",
        stringType : 'hex',
        count: 8},
    {field: "year",
        index: 17,
        expression: function (input) {
            return (input + 2010)
        },
        signed: false,
        type: "bits",
        bitPosition: 0,
        count: 4},
    {field: "month",
        index: 17,
        expression: null,
        signed: false,
        type: "bits",
        bitPosition: 4,
        count: 4},
    {field: "day",
        index: 18,
        expression: null,
        type : "byteInt",
        signed : false,
        count: 1},
    {field: "hour",
        index: 19,
        expression: null,
        type : "byteInt",
        signed : false,
        count: 1},
    {field: "minutes",
        index: 20,
        expression: null,
        type : "byteInt",
        signed : false,
        count: 1},
    {field: "latitude",
        index : 21,
        expression:function(input){
            var dd =  Math.round(input/100000);
            var mmsss =  (input%100000)/1000;
            return dd + (mmsss/60)},
        type : "integer",
        LEorBE : "BE",
        signed : true,
        count: 4},
    {field: "longitude",
        index : 25,
        expression: function(input){
            var dd =  Math.round(input/100000);
            var mmsss =  (input%100000)/1000;
            return dd + (mmsss/60)},
        LEorBE : "BE",
        type : "integer",
        signed : true,
        count: 4},
    {field: "Altitude",
        index : 29,
        expression: null,
        signed : true,
        type : "smallInt",
        LEorBE : "BE",
        count: 2},
    {field: "heading",
        index : 31,
        type : "smallInt",
        expression: null,
        signed : false,
        LEorBE : "BE",
        count: 2},
    {field: "velocity",
        index : 33,
        expression: null,
        type : "byteInt",
        signed : false,
        count: 1},
    {field: "powerInputVoltage",
        index : 34,
        expression: function(input){
            return (input*0.1)},
        type : "byteInt",
        signed : false,
        count: 1},
    {field: "gpioStatusAndDirection",
        index : 35,
        expression: null,
        type : "byteInt",
        signed : false,
        count: 1},
    {field: "timeSeconds",
        index : 36,
        expression: null,
        type : "byteInt",
        signed : true,
        count: 1},
    {field: "rssi",
        index : 37,
        expression: null,
        type : "byteInt",
        signed : false,
        count: 1},
    {field: "sequenceCounter",
        index : 38,
        expression: null,
        type : "byteInt",
        signed : false,
        count: 1},
    {field: "versionID",
        index : 39,
        expression: null,
        type : "byteInt",
        signed : false,
        count: 1},
    {field: "lockFlag",
        index: 40,
        expression: null,
        signed: false,
        type: "bits",
        bitPosition: 0,
        count: 1},
    {field: "oldFlag",
        index: 40,
        expression: null,
        signed: false,
        type: "bits",
        bitPosition: 1,
        count: 1},
    {field: "reservedBits",
        index: 40,
        expression: null,
        signed: false,
        type: "bits",
        bitPosition: 2,
        count: 2},
    {field: "numberOfVisibleSatellites",
        index: 40,
        expression: null,
        signed: false,
        type: "bits",
        bitPosition: 4,
        count: 4},
    {field: "virtualOdometer",
        index : 41,
        expression: null,
        type : "integer",
        LEorBE : "BE",
        signed : false,
        count: 4}
];
