// decode-engine/specs/moog.js
module.exports = [
    {field: "log",
        fieldType : 'others',
        expression: null
    },
    {field: "report_date",
        fieldType : 'log',
        index: 0,
        expression: null
    },
    {field: "report_time",
        fieldType : 'log',
        index: 1,
        expression: null
    },
    {field: "latitude",
        fieldType : 'log',
        index: 2,
        expression: function(input){
            return Number(input)}
    },
    {field: "lat_n_s",
        fieldType : 'log',
        index: 3,
        expression: null},
    {field: "longitude",
        fieldType : 'log',
        index: 4,
        expression: function(input){
            return Number(input)}
    },
    {field: "lng_e_w",
        fieldType : 'log',
        index: 5,
        expression: null
    },
    {field: "altitude",
        fieldType : 'log',
        index: 6,
        expression: function(input){
            return Number(input)}
    },
    {field: "speed",
        index: 7,
        fieldType : 'log',
        expression: function(input){
            return Number(input)}
    },
    {field: "heading",
        index : 8,
        fieldType : 'log',
        expression: function(input){
            return Number(input)}},
    {field: "num_satilites",
        index : 9,
        fieldType : 'log',
        expression: function(input){
            return Number(input)}},
    {field: "battery",
        index : 10,
        fieldType : 'log',
        expression: function(input){
            return Number(input)}
    },
    {field: "supply",
        index : 11,
        fieldType : 'log',
        expression: function(input){
            return Number(input)}
    },
    {field: "accel_x",
        index : 12,
        fieldType : 'log',
        expression: function(input){
            return Number(input)}
    },
    {field: "accel_y",
        index : 13,
        fieldType : 'log',
        expression: function(input){
            return Number(input)}
    },
    {field: "accel_z",
        index : 14,
        fieldType : 'log',
        expression: function(input){
            return Number(input)}
    },
    {field: "pressure",
        index : 15,
        fieldType : 'log',
        expression: function(input){
            return Number(input)}
    },
    {field: "humidity",
        index : 16,
        fieldType : 'log',
        expression: function(input){
            return Number(input)}
    },
    {field: "temperature",
        index : 17,
        fieldType : 'log',
        expression: function(input){
            return input/100}
    },
    {field: "light",
        index : 18,
        fieldType : 'log',
        expression: function(input){
            return Number(input)}
    },
    {field: "uid",
        index : 19,
        fieldType : 'log',
        expression:function(input){
            return input.substring(2)}
    },
    {field: "cell",
        fieldType : 'others',
        expression: null
    },
    {field: "mcc",
        index : 0,
        fieldType : 'cell',
        expression: function(input){
            return Number(input)}
    },
    {field: "mnc",
        index : 1,
        fieldType : 'cell',
        expression: function(input){
            return Number(input)}
    },
    {field: "lac",
        index : 2,
        fieldType : 'cell',
        expression: function(input){
            return Number(input)}
    },
    {field: "cid",
        index : 3,
        fieldType : 'cell',
        expression: function(input){
            return Number(input)}
    },
    {field: "cell_rssi",
        fieldType : 'others',
        expression: function(input){
            return Number(input)}
    },
    {field: "gps_hacc",
        fieldType : 'others',
        expression: function(input){
            return Number(input)}
    },
    {field: "gps_ttf",
        fieldType : 'others',
        expression: function(input){
            return Number(input)}
    },
    {field: "gps_fix",
        fieldType : 'others',
        expression: function(input){
            return Number(input)}
    },
    {field: "bootno",
        fieldType : 'others',
        expression: function(input){
            return Number(input)}
    },
    {field: "seqno",
        fieldType : 'others',
        expression: function(input){
            return Number(input)}
    },
    {field: "alerts_mask",
        fieldType : 'others',
        expression: null
    },
    {field: "v",
        fieldType : 'others',
        expression: null
    },
    {field: "fwd",
        fieldType : 'others',
        expression: function(input){
            return Number(input)}
    },
    {field: "acstate",
        fieldType : 'others',
        expression: null
    },
    {field: "acevt",
        fieldType : 'others',
        expression: function(input){
            return Number(input)}
    },
    {field: "motion",
        fieldType : 'others',
        expression: function(input){
            return Number(input)}
    },
    {field: "still_time",
        fieldType : 'others',
        expression: function(input){
            return Number(input)}
    },
    {field: "temp_h",
        fieldType : 'others',
        expression: null
    },
    {field: "temp_a",
        fieldType : 'others',
        expression: null
    },
    {field: "temp_b",
        fieldType : 'others',
        expression: null
    },
    {field: "lux",
        fieldType : 'others',
        expression: null
    },
    {field: "TYPE",
        fieldType : 'others',
        expression: null
    },
    {field: "sim",
        fieldType : 'others',
        expression: function(input){
            return Number(input)}
    },
    {field: "phone",
        fieldType : 'others',
        expression: function(input){
            return Number(input)}
    },
    {field: "sw_ver",
        fieldType : 'others',
        expression: null
    },
    {field: "hw_ver",
        fieldType : 'others',
        expression: null
    },
    {field: "fw_ver",
        fieldType : 'others',
        expression: null
    },
    {field: "mac",
        fieldType : 'others',
        expression: null
    },
    {field: "shock",
        fieldType : 'others',
        expression: null
    },
    {field: "fwd",
        fieldType : 'others',
        expression: null
    },
    {field: "battery",
        index: 0,
        expression: null,
        fieldType : 'alerts',
        signed: false,
        type: "bits",
        fromBytes : 4,
        LEorBE : "BE",
        bitPosition: 31,
        count: 1},
    {field: "temperature",
        index: 0,
        expression: null,
        fieldType : 'alerts',
        signed: false,
        type: "bits",
        bitPosition: 30,
        fromBytes : 4,
        LEorBE : "BE",
        count: 1},
    {field: "humidity",
        index: 0,
        expression: null,
        fieldType : 'alerts',
        signed: false,
        type: "bits",
        bitPosition: 29,
        fromBytes : 4,
        LEorBE : "BE",
        count: 1},
    {field: "pressure",
        index: 0,
        expression: null,
        fieldType : 'alerts',
        signed: false,
        type: "bits",
        fromBytes : 4,
        LEorBE : "BE",
        bitPosition: 28,
        count: 1},
    {field: "accelerationx",
        index: 0,
        expression: null,
        fieldType : 'alerts',
        signed: false,
        type: "bits",
        fromBytes : 4,
        LEorBE : "BE",
        bitPosition: 27,
        count: 1},
    {field: "accelerationy",
        index: 0,
        expression: null,
        fieldType : 'alerts',
        signed: false,
        type: "bits",
        LEorBE : "BE",
        fromBytes : 4,
        bitPosition: 26,
        count: 1},
    {field: "accelerationz",
        index: 0,
        expression: null,
        fieldType : 'alerts',
        signed: false,
        type: "bits",
        bitPosition: 25,
        LEorBE : "BE",
        fromBytes : 4,
        count: 1},
    {field: "light",
        index: 0,
        expression: null,
        fieldType : 'alerts',
        signed: false,
        type: "bits",
        bitPosition: 20,
        LEorBE : "BE",
        fromBytes : 4,
        count: 1},
    {field: "shock",
        index: 0,
        expression: null,
        signed: false,
        fieldType : 'alerts',
        type: "bits",
        bitPosition: 19,
        LEorBE : "BE",
        fromBytes : 4,
        count: 1},
    {field: "fall",
        index: 0,
        expression: null,
        fieldType : 'alerts',
        signed: false,
        type: "bits",
        LEorBE : "BE",
        bitPosition: 18,
        fromBytes : 4,
        count: 1},
    {field: "tilt",
        index: 0,
        expression: null,
        fieldType : 'alerts',
        signed: false,
        type: "bits",
        bitPosition: 17,
        LEorBE : "BE",
        fromBytes : 4,
        count: 1},
    {field: "probe",
        index: 0,
        expression: null,
        fieldType : 'alerts',
        signed: false,
        type: "bits",
        LEorBE : "BE",
        bitPosition: 15,
        fromBytes : 4,
        count: 1}
];
